modules/core/src/main/java/org/apache/ignite/internal/processors/affinity/GridAffinityAssignmentCache.java

```diff
@@ -978,20 +968,26 @@ private void onHistoryAdded(
 
             Iterator<HistoryAffinityAssignment> it = affCache.values().iterator();
 
+            AffinityTopologyVersion lastAffChangeTopVer =
+                ctx.cache().context().exchange().lastAffinityChangedTopologyVersion(head.get().topologyVersion());
+
             while (it.hasNext()) {
                 HistoryAffinityAssignment aff0 = it.next();
 
-                if (aff0.requiresHistoryCleanup()) {
-                    // We can stop cleanup only on non-shallow item.
-                    // Keeping part of shallow items chain if corresponding real item is missing makes no sense.
-                    if (!shouldContinueCleanup(nonShallowSize, totalSize)) {
-                        nonShallowHistSize.getAndAdd(nonShallowSize - initNonShallowSize);
+                if (!shouldContinueCleanup(nonShallowSize, totalSize)) {
```

Comment:
```
            int initNonShallowSize = nonShallowSize;
```

This variable not required and always equals to `nonShallowHistSize.get()`.
Can be removed.

This line 
```
nonShallowHistSize.getAndAdd(nonShallowSize - initNonShallowSize);
```
can be replaced with
```
nonShallowHistSize.set(nonShallowSize);
```

Reply:
I think the main idea here is thread-safety. There are a lot of methods, which call `initialize` method, that calls `onHistoryAdded`, perhaps this method can be called concurrently. When we atomically substract from `nonShallowHistSize` it's thread safe, when we use `nonShallowHistSize.set` it's not thread safe.